File: src-tauri/src/library.rs
use crate::epub_meta;
use crate::models::{BookMeta, BookSummary, LibraryIndex, NoteSearchHit};
use crate::notes;
use chrono::Utc;
use sha2::{Digest, Sha256};
use std::fs;
use std::io::{Read, Write};
use std::path::{Path, PathBuf};
use thiserror::Error;
use uuid::Uuid;

#[derive(Debug, Error)]
pub enum LibraryError {
    #[error("io error: {0}")]
    Io(#[from] std::io::Error),
    #[error("json error: {0}")]
    Json(#[from] serde_json::Error),
    #[error("epub error: {0}")]
    Epub(#[from] epub_meta::EpubError),
    #[error("notes error: {0}")]
    Notes(#[from] notes::NotesError),
    #[error("{0}")]
    Other(String),
}

pub struct Library {
    root: PathBuf,
}

impl Library {
    pub fn open(root: PathBuf) -> Result<Self, LibraryError> {
        fs::create_dir_all(&root)?;
        fs::create_dir_all(root.join("books"))?;
        Ok(Self { root })
    }

    pub fn set_root(&mut self, root: PathBuf) -> Result<(), LibraryError> {
        fs::create_dir_all(&root)?;
        fs::create_dir_all(root.join("books"))?;
        self.root = root;
        Ok(())
    }

    pub fn book_dir(&self, book_id: &str) -> PathBuf {
        self.root.join("books").join(book_id)
    }

    pub fn list_books(&self) -> Result<Vec<BookSummary>, LibraryError> {
        let books_dir = self.root.join("books");
        let mut summaries = Vec::new();

        if !books_dir.exists() {
            return Ok(summaries);
        }

        for entry in fs::read_dir(&books_dir)? {
            let entry = entry?;
            if entry.file_name().to_string_lossy().starts_with('.') {
                continue;
            }
            if !entry.file_type()?.is_dir() {
                continue;
            }
            let meta_path = entry.path().join("meta.json");
            if !meta_path.exists() {
                continue;
            }
            let meta: BookMeta = serde_json::from_str(&fs::read_to_string(meta_path)?)?;
            let notes_count = notes::count_notes(&entry.path())?;
            summaries.push(BookSummary {
                id: meta.id,
                title: meta.title,
                author: meta.author,
                added_at: meta.added_at,
                chapter_count: meta.chapters.len(),
                notes_count,
            });
        }

        summaries.sort_by_key(|summary| std::cmp::Reverse(summary.added_at));
        self.write_index(&summaries)?;
        Ok(summaries)
    }

    pub fn get_book(&self, book_id: &str) -> Result<BookMeta, LibraryError> {
        let meta_path = self.book_dir(book_id).join("meta.json");
        if !meta_path.exists() {
            return Err(LibraryError::Other(format!("book not found: {book_id}")));
        }
        Ok(serde_json::from_str(&fs::read_to_string(meta_path)?)?)
    }

    pub fn read_epub_bytes(&self, book_id: &str) -> Result<Vec<u8>, LibraryError> {
        let epub_path = self.book_dir(book_id).join("source.epub");
        if !epub_path.exists() {
            return Err(LibraryError::Other(format!(
                "epub missing for book: {book_id}"
            )));
        }
        Ok(fs::read(epub_path)?)
    }

    pub fn import_epub_with_progress<F>(
        &self,
        source_path: PathBuf,
        mut on_progress: F,
    ) -> Result<BookMeta, LibraryError>
    where
        F: FnMut(u8, &'static str),
    {
        on_progress(0, "preparing");
        if !source_path.exists() {
            return Err(LibraryError::Other("source file does not exist".into()));
        }

        on_progress(5, "reading-metadata");
        let info = epub_meta::parse_epub(&source_path)?;
        on_progress(30, "hashing");
        let book_id = hash_file_with_progress(&source_path, |processed, total| {
            on_progress(progress_between(30, 50, processed, total), "hashing");
        })?;
        let final_book_dir = self.book_dir(&book_id);

        if final_book_dir.join("meta.json").is_file()
            && final_book_dir.join("source.epub").is_file()
        {
            let existing = self.get_book(&book_id)?;
            on_progress(100, "already-imported");
            return Ok(existing);
        }

        let staging = ImportStaging::new(
            self.root
                .join("books")
                .join(format!(".{book_id}.importing-{}", Uuid::new_v4())),
        );

        on_progress(55, "copying");
        fs::create_dir_all(staging.path.join("notes/chapters"))?;

        let source_filename = source_path
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or("book.epub")
            .to_string();

        copy_file_with_progress(
            &source_path,
            &staging.path.join("source.epub"),
            |processed, total| {
                on_progress(progress_between(55, 94, processed, total), "copying");
            },
        )?;

        let meta = BookMeta {
            id: book_id.clone(),
            title: info.title,
            author: info.author,
            language: info.language,
            added_at: Utc::now(),
            source_filename,
            chapters: info.chapters,
        };

        on_progress(95, "saving");
        let meta_raw = serde_json::to_string_pretty(&meta)?;
        fs::write(staging.path.join("meta.json"), meta_raw)?;

        on_progress(97, "saving");
        notes::write_empty_index(&staging.path)?;
        on_progress(99, "saving");
        self.write_book_readme(&staging.path, &meta)?;

        if final_book_dir.exists() {
            // A previous interrupted import can leave an incomplete final directory.
            // Remove it only after the source has been copied into staging.
            fs::remove_dir_all(&final_book_dir)?;
        }
        staging.commit(&final_book_dir)?;
        on_progress(100, "complete");
        Ok(meta)
    }

    pub fn remove_book(&self, book_id: &str) -> Result<(), LibraryError> {
        let book_dir = self.book_dir(book_id);
        if book_dir.exists() {
            fs::remove_dir_all(book_dir)?;
        }
        Ok(())
    }

    pub fn search_notes(&self, query: &str) -> Result<Vec<NoteSearchHit>, LibraryError> {
        Ok(notes::search_notes(&self.root, query)?)
    }

    fn write_index(&self, summaries: &[BookSummary]) -> Result<(), LibraryError> {
        let index = LibraryIndex {
            books: summaries.to_vec(),
        };
        let raw = serde_json::to_string_pretty(&index)?;
        fs::write(self.root.join("index.json"), raw)?;
        Ok(())
    }

    fn write_book_readme(&self, book_dir: &Path, meta: &BookMeta) -> Result<(), LibraryError> {
        let readme = format!(
            "# {}\n\n\
             Author: {}\n\
             Book ID: `{}`\n\n\
             ## Notes layout\n\n\
             Chapter summaries and annotations live in `notes/chapters/` as Markdown files \
             with YAML frontmatter. Each file is self-contained and agent-friendly.\n\n\
             - `meta.json` — book metadata and chapter spine\n\
             - `notes/_index.json` — machine-readable note index\n\
             - `notes/chapters/*.md` — one file per chapter note\n\
             - `source.epub` — imported EPUB copy\n",
            meta.title, meta.author, meta.id
        );
        fs::write(book_dir.join("README.md"), readme)?;
        Ok(())
    }
}

struct ImportStaging {
    path: PathBuf,
    committed: bool,
}

impl ImportStaging {
    fn new(path: PathBuf) -> Self {
        Self {
            path,
            committed: false,
        }
    }

    fn commit(mut self, destination: &Path) -> Result<(), std::io::Error> {
        fs::rename(&self.path, destination)?;
        self.committed = true;
        Ok(())
    }
}

impl Drop for ImportStaging {
    fn drop(&mut self) {
        if !self.committed {
            let _ = fs::remove_dir_all(&self.path);
        }
    }
}

fn hash_file_with_progress<F>(path: &Path, mut on_progress: F) -> Result<String, LibraryError>
where
    F: FnMut(u64, u64),
{
    let mut file = fs::File::open(path)?;
    let total = file.metadata()?.len();
    let mut hasher = Sha256::new();
    let mut buffer = [0u8; 8192];
    let mut processed = 0;
    on_progress(processed, total);
    loop {
        let read = file.read(&mut buffer)?;
        if read == 0 {
            break;
        }
        hasher.update(&buffer[..read]);
        processed += read as u64;
        on_progress(processed, total);
    }
    Ok(hex::encode(&hasher.finalize()[..12]))
}

fn copy_file_with_progress<F>(
    source: &Path,
    destination: &Path,
    mut on_progress: F,
) -> Result<(), std::io::Error>
where
    F: FnMut(u64, u64),
{
    let mut source_file = fs::File::open(source)?;
    let total = source_file.metadata()?.len();
    let mut destination_file = fs::File::create(destination)?;
    let mut buffer = [0u8; 64 * 1024];
    let mut processed = 0;
    on_progress(processed, total);

    loop {
        let read = source_file.read(&mut buffer)?;
        if read == 0 {
            break;
        }
        destination_file.write_all(&buffer[..read])?;
        processed += read as u64;
        on_progress(processed, total);
    }

    destination_file.flush()?;
    Ok(())
}

fn progress_between(start: u8, end: u8, processed: u64, total: u64) -> u8 {
    if total == 0 {
        return end;
    }

    let range = u128::from(end.saturating_sub(start));
    let completed = u128::from(processed.min(total));
    let total = u128::from(total);
    start.saturating_add((range * completed / total) as u8)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::test_fixtures::write_sample_epub;

    #[test]
    fn import_list_get_remove_roundtrip() {
        let tmp = tempfile::tempdir().unwrap();
        let library = Library::open(tmp.path().join("library")).unwrap();
        let epub = write_sample_epub(tmp.path(), "sample.epub");

        let meta = library
            .import_epub_with_progress(epub.clone(), |_, _| {})
            .unwrap();
        assert_eq!(meta.title, "Sample Book");
        assert_eq!(meta.chapters.len(), 2);
        assert!(library.book_dir(&meta.id).join("source.epub").exists());
        assert!(library
            .book_dir(&meta.id)
            .join("notes/_index.json")
            .exists());

        let again = library.import_epub_with_progress(epub, |_, _| {}).unwrap();
        assert_eq!(again.id, meta.id);

        let listed = library.list_books().unwrap();
        assert_eq!(listed.len(), 1);
        assert_eq!(listed[0].id, meta.id);
        assert_eq!(listed[0].notes_count, 0);
        assert!(tmp.path().join("library/index.json").exists());

        let bytes = library.read_epub_bytes(&meta.id).unwrap();
        assert!(!bytes.is_empty());

        library.remove_book(&meta.id).unwrap();
        assert!(library.list_books().unwrap().is_empty());
        assert!(!library.book_dir(&meta.id).exists());
    }

    #[test]
    fn import_reports_monotonic_progress() {
        let tmp = tempfile::tempdir().unwrap();
        let library = Library::open(tmp.path().join("library")).unwrap();
        let epub = write_sample_epub(tmp.path(), "sample.epub");
        let mut updates = Vec::new();

        library
            .import_epub_with_progress(epub, |percent, stage| updates.push((percent, stage)))
            .unwrap();

        assert_eq!(updates.first(), Some(&(0, "preparing")));
        assert_eq!(updates.last(), Some(&(100, "complete")));
        assert!(updates.iter().any(|(_, stage)| *stage == "hashing"));
        assert!(updates.iter().any(|(_, stage)| *stage == "copying"));
        assert!(updates.windows(2).all(|pair| pair[1].0 >= pair[0].0));
    }

    #[test]
    fn failed_import_can_be_retried_without_a_partial_book() {
        let tmp = tempfile::tempdir().unwrap();
        let library = Library::open(tmp.path().join("library")).unwrap();
        let epub = write_sample_epub(tmp.path(), "sample.epub");
        let original = fs::read(&epub).unwrap();
        let mut failed_once = false;

        let first_attempt = library.import_epub_with_progress(epub.clone(), |_, stage| {
            if stage == "copying" && !failed_once {
                failed_once = true;
                fs::remove_file(&epub).unwrap();
            }
        });

        assert!(first_attempt.is_err());
        fs::write(&epub, original).unwrap();
        assert!(library.import_epub_with_progress(epub, |_, _| {}).is_ok());
    }

    #[test]
    fn repairing_incomplete_book_does_not_delete_its_source() {
        let tmp = tempfile::tempdir().unwrap();
        let library = Library::open(tmp.path().join("library")).unwrap();
        let original_epub = write_sample_epub(tmp.path(), "sample.epub");
        let book_id = hash_file_with_progress(&original_epub, |_, _| {}).unwrap();
        let partial_dir = library.book_dir(&book_id);
        fs::create_dir_all(&partial_dir).unwrap();
        fs::copy(&original_epub, partial_dir.join("source.epub")).unwrap();

        assert!(library
            .import_epub_with_progress(partial_dir.join("source.epub"), |_, _| {})
            .is_ok());
    }

    #[test]
    fn list_books_ignores_interrupted_import_staging_directories() {
        let tmp = tempfile::tempdir().unwrap();
        let library = Library::open(tmp.path().join("library")).unwrap();
        let epub = write_sample_epub(tmp.path(), "sample.epub");
        let meta = library.import_epub_with_progress(epub, |_, _| {}).unwrap();
        let staging_dir = library
            .book_dir(&meta.id)
            .parent()
            .unwrap()
            .join(format!(".{}.importing-crashed", meta.id));
        fs::create_dir_all(&staging_dir).unwrap();
        fs::copy(
            library.book_dir(&meta.id).join("meta.json"),
            staging_dir.join("meta.json"),
        )
        .unwrap();

        let listed = library.list_books().unwrap();
        assert_eq!(listed.len(), 1);
        assert_eq!(listed[0].id, meta.id);
    }

    #[test]
    fn failed_duplicate_import_does_not_report_completion() {
        let tmp = tempfile::tempdir().unwrap();
        let library = Library::open(tmp.path().join("library")).unwrap();
        let epub = write_sample_epub(tmp.path(), "sample.epub");
        let meta = library
            .import_epub_with_progress(epub.clone(), |_, _| {})
            .unwrap();
        fs::write(library.book_dir(&meta.id).join("meta.json"), b"not json").unwrap();
        let mut updates = Vec::new();

        assert!(library
            .import_epub_with_progress(epub, |percent, stage| updates.push((percent, stage)))
            .is_err());
        assert!(!updates.iter().any(|(percent, _)| *percent == 100));
    }

    #[test]
    fn reimport_repairs_a_missing_source_file() {
        let tmp = tempfile::tempdir().unwrap();
        let library = Library::open(tmp.path().join("library")).unwrap();
        let epub = write_sample_epub(tmp.path(), "sample.epub");
        let meta = library
            .import_epub_with_progress(epub.clone(), |_, _| {})
            .unwrap();
        fs::remove_file(library.book_dir(&meta.id).join("source.epub")).unwrap();

        library.import_epub_with_progress(epub, |_, _| {}).unwrap();
        assert!(library.book_dir(&meta.id).join("source.epub").is_file());
    }

    #[test]
    fn set_root_switches_active_library() {
        let tmp = tempfile::tempdir().unwrap();
        let root_a = tmp.path().join("a");
        let root_b = tmp.path().join("b");
        let epub = write_sample_epub(tmp.path(), "sample.epub");

        let mut library = Library::open(root_a).unwrap();
        let meta = library.import_epub_with_progress(epub, |_, _| {}).unwrap();
        assert_eq!(library.list_books().unwrap().len(), 1);

        library.set_root(root_b).unwrap();
        assert!(library.list_books().unwrap().is_empty());
        assert!(!library.book_dir(&meta.id).exists());
    }
}

File: apple/Sources/MarginsCore/Library.swift
import CryptoKit
import Foundation

// The library tree: import, catalog, reading positions, bookmarks, removal
// (docs/storage.md). A book directory is named by a content hash of its
// EPUB, so importing the same file twice is idempotent, and it is assembled
// in a hidden staging directory and renamed into place, so an interrupted
// import can never appear as a book.

public final class Library {
    /// Schema version of `BookMeta.chapters`. Bumped whenever chapter
    /// metadata gains information the parser can now recover from the source
    /// EPUB (v1: TOC-derived titles and start fragments; v2: matter
    /// classification, outline levels, and every TOC entry per file); the
    /// library scan re-parses any book below it. Chapter keys are
    /// spine-derived and stay stable across re-parses, so notes keep
    /// resolving.
    public static let chaptersVersion = 2

    public private(set) var root: String
    private let search = SearchEngine()

    public init(root: String) throws {
        self.root = root
        try Self.prepare(root)
    }

    public func setRoot(_ root: String) throws {
        try Self.prepare(root)
        self.root = root
        // The cached index refers to the previous root; the next query
        // rebuilds it against the new one.
        search.clear()
    }

    private static func prepare(_ root: String) throws {
        try Files.createDirectory(root)
        try Files.createDirectory(root.appendingPathComponent("books"))
    }

    public func bookDir(_ bookID: String) -> String {
        root.appendingPathComponent("books").appendingPathComponent(bookID)
    }

    // MARK: The catalog

    public func listBooks() throws -> [BookSummary] {
        let booksDir = root.appendingPathComponent("books")
        guard Files.exists(booksDir) else { return [] }

        var summaries: [BookSummary] = []
        for path in try Files.contents(ofDirectory: booksDir) {
            let name = (path as NSString).lastPathComponent
            guard !name.hasPrefix("."), Files.isDirectory(path) else { continue }
            let metaPath = path.appendingPathComponent("meta.json")
            guard FileStore.exists(metaPath) else { continue }

            var meta = try readMeta(at: metaPath)
            backfillChapters(bookDir: path, meta: &meta)
            summaries.append(
                BookSummary(
                    id: meta.id,
                    title: meta.title,
                    author: meta.author,
                    addedAt: meta.addedAt,
                    chapterCount: meta.chapters.count,
                    notesCount: try Notes.countNotes(bookDir: path),
                    cover: meta.cover ?? backfillCover(bookDir: path, meta: meta),
                    progressPercent: readPosition(bookID: meta.id)?.percent
                )
            )
        }

        summaries.sort { $0.addedAt > $1.addedAt }
        try writeIndex(summaries)
        return summaries.map(resolvingCover)
    }

    public func getBook(id: String) throws -> BookMeta {
        let metaPath = bookDir(id).appendingPathComponent("meta.json")
        guard FileStore.exists(metaPath) else {
            throw CoreError.library("book not found: \(id)")
        }
        var meta = try readMeta(at: metaPath)
        meta.progressPercent = readPosition(bookID: id)?.percent
        return resolvingCover(meta)
    }

    public func removeBook(id: String) throws {
        let dir = bookDir(id)
        if Files.exists(dir) { try Files.remove(dir) }
    }

    // MARK: Reading position

    /// The book's saved reading position, or `nil` when it was never opened
    /// or the position file is missing/corrupt (a corrupt file must not
    /// break the library — the reader falls back to chapter 1).
    public func readPosition(bookID: String) -> ReadingPosition? {
        let path = bookDir(bookID).appendingPathComponent("position.json")
        guard let data = try? FileStore.readData(path) else { return nil }
        return try? MarginsJSON.decode(ReadingPosition.self, from: data)
    }

    /// Saves the book's reading position. Percent is clamped to 0–100 and
    /// the save time is stamped here; any caller value is advisory.
    public func writePosition(bookID: String, position: ReadingPosition) throws {
        let dir = bookDir(bookID)
        guard Files.isDirectory(dir) else {
            throw CoreError.library("book not found: \(bookID)")
        }
        var position = position
        position.percent = min(max(position.percent, 0), 100)
        position.updatedAt = RFC3339.now()
        try FileStore.writeData(
            MarginsJSON.encode(position), to: dir.appendingPathComponent("position.json")
        )
    }

    // MARK: Bookmarks

    /// Named location pins for the book. A missing or corrupt file is an
    /// empty list — same posture as a missing `position.json`.
    public func readBookmarks(bookID: String) -> [Bookmark] {
        let path = bookDir(bookID).appendingPathComponent("bookmarks.json")
        guard let data = try? FileStore.readData(path) else { return [] }
        guard let file = try? MarginsJSON.decode(BookmarkFile.self, from: data) else {
            return []
        }
        return Bookmark.sortedForDisplay(file.bookmarks)
    }

    /// Drops a pin at `position`. Empty `label` is untitled. Percent is
    /// clamped; the core stamps created/updated.
    @discardableResult
    public func addBookmark(
        bookID: String, label: String, position: ReadingPosition
    ) throws -> Bookmark {
        var bookmarks = readBookmarks(bookID: bookID)
        let taken = Set(bookmarks.map(\.id))
        var id = CoreID.newID()
        while taken.contains(id) { id = CoreID.newID() }
        let now = RFC3339.now()
        let bookmark = Bookmark(
            id: id,
            label: label.trimmingCharacters(in: .whitespacesAndNewlines),
            chapterKey: position.chapterKey,
            epubCfi: position.epubCfi.flatMap { $0.isEmpty ? nil : $0 },
            percent: min(max(position.percent, 0), 100),
            createdAt: now,
            updatedAt: now
        )
        bookmarks.append(bookmark)
        try writeBookmarks(bookID: bookID, bookmarks: bookmarks)
        return bookmark
    }

    /// Renames and/or restamps an existing pin. `nil` fields stay as they are.
    @discardableResult
    public func updateBookmark(
        bookID: String, id: String, label: String?, position: ReadingPosition?
    ) throws -> Bookmark {
        var bookmarks = readBookmarks(bookID: bookID)
        guard let index = bookmarks.firstIndex(where: { $0.id == id }) else {
            throw CoreError.library("bookmark not found: \(id)")
        }
        var bookmark = bookmarks[index]
        if let label {
            bookmark.label = label.trimmingCharacters(in: .whitespacesAndNewlines)
        }
        if let position {
            bookmark.chapterKey = position.chapterKey
            bookmark.epubCfi = position.epubCfi.flatMap { $0.isEmpty ? nil : $0 }
            bookmark.percent = min(max(position.percent, 0), 100)
        }
        bookmark.updatedAt = RFC3339.now()
        bookmarks[index] = bookmark
        try writeBookmarks(bookID: bookID, bookmarks: bookmarks)
        return bookmark
    }

    public func deleteBookmark(bookID: String, id: String) throws {
        var bookmarks = readBookmarks(bookID: bookID)
        let before = bookmarks.count
        bookmarks.removeAll { $0.id == id }
        guard bookmarks.count != before else {
            throw CoreError.library("bookmark not found: \(id)")
        }
        try writeBookmarks(bookID: bookID, bookmarks: bookmarks)
    }

    private func writeBookmarks(bookID: String, bookmarks: [Bookmark]) throws {
        let dir = bookDir(bookID)
        guard Files.isDirectory(dir) else {
            throw CoreError.library("book not found: \(bookID)")
        }
        let path = dir.appendingPathComponent("bookmarks.json")
        if bookmarks.isEmpty {
            if FileStore.exists(path) { try FileStore.remove(path) }
            return
        }
        try FileStore.writeData(
            MarginsJSON.encode(BookmarkFile(bookmarks: Bookmark.sortedForDisplay(bookmarks))),
            to: path
        )
    }

    public func readEpubBytes(bookID: String) throws -> Data {
        try Self.readEpubBytes(bookId: bookID, root: root)
    }

    /// Root-parameterized byte read. Exists for `CoreStore.readEpubBytesSync`,
    /// which the reader's scheme handler calls from WebKit threads — off the
    /// actor, so it carries the root with the call instead of reading the
    /// instance's.
    public static func readEpubBytes(bookId bookID: String, root: String) throws -> Data {
        let path = root.appendingPathComponent("books").appendingPathComponent(bookID)
            .appendingPathComponent("source.epub")
        guard Files.exists(path) else {
            throw CoreError.library("epub missing for book: \(bookID)")
        }
        return try Files.readData(path)
    }

    // MARK: Import

    /// Copies an EPUB into the library and records its metadata.
    ///
    /// `progress` is called with a percentage and a stage name; the import
    /// UIs show both. The work is assembled under `.{id}.importing-{uuid}`
    /// and renamed into place at the end, so a crash mid-import leaves a
    /// hidden directory the catalog ignores rather than a half-formed book.
    @discardableResult
    public func importEpub(
        atPath sourcePath: String,
        progress: (Int, String) -> Void = { _, _ in }
    ) throws -> BookMeta {
        progress(0, "preparing")
        guard Files.exists(sourcePath) else {
            throw CoreError.library("source file does not exist")
        }

        progress(5, "reading-metadata")
        let info = try EpubParser.parse(path: sourcePath)

        progress(30, "hashing")
        let bookID = try hashFile(sourcePath) { processed, total in
            progress(percentBetween(30, 50, processed, total), "hashing")
        }
        let finalDir = bookDir(bookID)

        if FileStore.isFile(finalDir.appendingPathComponent("meta.json")),
           Files.isFile(finalDir.appendingPathComponent("source.epub")) {
            let existing = try getBook(id: bookID)
            progress(100, "already-imported")
            return existing
        }

        let staging = root.appendingPathComponent("books")
            .appendingPathComponent(".\(bookID).importing-\(UUID().uuidString)")
        var committed = false
        defer { if !committed { try? Files.remove(staging) } }

        progress(55, "copying")
        try Files.createDirectory(staging.appendingPathComponent("notes/chapters"))
        try copyFile(
            from: sourcePath, to: staging.appendingPathComponent("source.epub")
        ) { processed, total in
            progress(percentBetween(55, 94, processed, total), "copying")
        }

        // A cover is a nice-to-have: a failed write must not fail the
        // import, so the meta records the name only on success.
        var coverName = info.cover.map { "cover.\($0.fileExtension)" }
        if let cover = info.cover, let name = coverName {
            do {
                try Files.writeData(cover.bytes, to: staging.appendingPathComponent(name))
            } catch {
                coverName = nil
            }
        }

        let meta = BookMeta(
            id: bookID,
            title: info.title,
            author: info.author,
            language: info.language,
            addedAt: RFC3339.now(),
            sourceFilename: (sourcePath as NSString).lastPathComponent,
            chapters: info.chapters,
            cover: coverName,
            chaptersVersion: Self.chaptersVersion
        )

        progress(95, "saving")
        try FileStore.writeData(
            MarginsJSON.encode(meta), to: staging.appendingPathComponent("meta.json")
        )
        progress(97, "saving")
        try Notes.writeEmptyIndex(bookDir: staging)
        progress(99, "saving")
        try writeBookReadme(bookDir: staging, meta: meta)

        if Files.exists(finalDir) {
            // A previous interrupted import can leave an incomplete final
            // directory. Remove it only after the source has been copied
            // into staging.
            try Files.remove(finalDir)
        }
        try Files.rename(staging, to: finalDir)
        committed = true

        progress(100, "complete")
        return resolvingCover(meta)
    }

    // MARK: Search

    public func searchNotes(query: String) -> [NoteSearchHit] {
        search.query(root: root, raw: query)
    }

    /// Updates the search index for one book in place after a save.
    public func refreshNoteIndex(bookID: String) {
        search.refreshBook(root: root, bookID: bookID)
    }

    // MARK: Backfills

    /// Re-derives chapter metadata for books imported under an older parser
    /// (`chaptersVersion` below the current one) by re-reading the retained
    /// `source.epub`. The spine is unchanged, so keys keep pointing at the
    /// same notes and `position.json`; only titles and start fragments
    /// improve. A missing or corrupt source leaves the book exactly as it
    /// was — a bad EPUB must never break the library scan.
    private func backfillChapters(bookDir: String, meta: inout BookMeta) {
        guard meta.chaptersVersion < Self.chaptersVersion else { return }
        let source = bookDir.appendingPathComponent("source.epub")
        guard Files.isFile(source), let info = try? EpubParser.parse(path: source) else { return }

        meta.chapters = info.chapters
        meta.chaptersVersion = Self.chaptersVersion
        try? FileStore.writeData(
            MarginsJSON.encode(meta), to: bookDir.appendingPathComponent("meta.json")
        )
    }

    /// One-shot cover backfill for books imported before covers were
    /// extracted: pulls the cover out of the retained `source.epub`, writes
    /// it into the book directory, and records the file name in `meta.json`.
    /// Books whose EPUB has no cover are re-probed on each scan (a cheap zip
    /// central-directory read at personal-library scale).
    private func backfillCover(bookDir: String, meta: BookMeta) -> String? {
        let source = bookDir.appendingPathComponent("source.epub")
        guard Files.isFile(source), let cover = EpubParser.extractCover(path: source) else {
            return nil
        }
        let name = "cover.\(cover.fileExtension)"
        guard (try? Files.writeData(cover.bytes, to: bookDir.appendingPathComponent(name))) != nil
        else { return nil }

        var updated = meta
        updated.cover = name
        try? FileStore.writeData(
            MarginsJSON.encode(updated), to: bookDir.appendingPathComponent("meta.json")
        )
        return name
    }

    // MARK: Writing

    private func writeIndex(_ summaries: [BookSummary]) throws {
        try Files.writeData(
            MarginsJSON.encode(LibraryIndex(books: summaries)),
            to: root.appendingPathComponent("index.json")
        )
    }

    private func writeBookReadme(bookDir: String, meta: BookMeta) throws {
        let readme = """
        # \(meta.title)

        Author: \(meta.author)
        Book ID: `\(meta.id)`

        ## Notes layout

        Chapter summaries and annotations live in `notes/chapters/` as \
        Markdown files with YAML frontmatter. Each file is self-contained \
        and agent-friendly.

        - `meta.json` — book metadata and chapter spine
        - `notes/_index.json` — machine-readable note index
        - `notes/chapters/*.md` — one file per chapter note
        - `source.epub` — imported EPUB copy

        """
        try Files.write(readme, to: bookDir.appendingPathComponent("README.md"))
    }

    private func readMeta(at path: String) throws -> BookMeta {
        do {
            return try MarginsJSON.decode(BookMeta.self, from: FileStore.readData(path))
        } catch let error as CoreError {
            throw error
        } catch {
            throw CoreError.library("json error: \(error.localizedDescription)")
        }
    }

    /// Fills in the absolute cover path callers display from. The stored
    /// `cover` stays a bare file name so the tree is portable.
    private func resolvingCover(_ meta: BookMeta) -> BookMeta {
        var meta = meta
        meta.coverPath = meta.cover.map { bookDir(meta.id).appendingPathComponent($0) }
        return meta
    }

    private func resolvingCover(_ summary: BookSummary) -> BookSummary {
        var summary = summary
        summary.coverPath = summary.cover.map { bookDir(summary.id).appendingPathComponent($0) }
        return summary
    }
}

// MARK: - File work

/// The book id: the first 12 bytes of the EPUB's SHA-256, hex-encoded — 24
/// characters, enough to make a collision in a personal library impossible
/// and short enough to type.
func hashFile(_ path: String, progress: (Int64, Int64) -> Void) throws -> String {
    guard let handle = FileHandle(forReadingAtPath: path) else {
        throw CoreError.io("io error: could not read \(path)")
    }
    defer { try? handle.close() }

    let total = fileSize(path)
    var hasher = SHA256()
    var processed: Int64 = 0
    progress(processed, total)

    while let chunk = try? handle.read(upToCount: 8192), !chunk.isEmpty {
        hasher.update(data: chunk)
        processed += Int64(chunk.count)
        progress(processed, total)
    }
    return hasher.finalize().prefix(12).map { String(format: "%02x", $0) }.joined()
}

/// Streams a copy so the import UI can show real progress on a large book.
func copyFile(from source: String, to destination: String, progress: (Int64, Int64) -> Void) throws {
    guard let reader = FileHandle(forReadingAtPath: source) else {
        throw CoreError.io("io error: could not read \(source)")
    }
    defer { try? reader.close() }

    guard FileManager.default.createFile(atPath: destination, contents: nil),
          let writer = FileHandle(forWritingAtPath: destination)
    else {
        throw CoreError.io("io error: could not write \(destination)")
    }
    defer { try? writer.close() }

    let total = fileSize(source)
    var processed: Int64 = 0
    progress(processed, total)

    while let chunk = try? reader.read(upToCount: 64 * 1024), !chunk.isEmpty {
        do {
            try writer.write(contentsOf: chunk)
        } catch {
            throw CoreError.io("io error: could not write \(destination)")
        }
        processed += Int64(chunk.count)
        progress(processed, total)
    }
}

private func fileSize(_ path: String) -> Int64 {
    let attributes = try? FileManager.default.attributesOfItem(atPath: path)
    return (attributes?[.size] as? NSNumber)?.int64Value ?? 0
}

/// Maps byte progress onto a slice of the overall percentage.
func percentBetween(_ start: Int, _ end: Int, _ processed: Int64, _ total: Int64) -> Int {
    guard total > 0 else { return end }
    let range = max(end - start, 0)
    return start + Int(Int64(range) * min(processed, total) / total)
}
